WIP Sparklines

ghronkrepps opened the pull request:
#### What's this PR do?
Add basic Sparkline support.

#### Where should the reviewer start?
The Sparklines folder

#### How should this be manually tested?
See the Sparklines example

#### Any background context you want to provide?
https://github.com/ClosedXML/ClosedXML/issues/454

#### Screenshots (if appropriate)
#### Questions:
- Is there a blog post? No
- Does the knowledge base need an update? Yes
- Does this add new (C#) dependencies? No

- [X] C# Code Review: @csreviewer
- [ ] Test Automation Review: @csreviewer

Depends on #949

vbjay commented:
You need to rebase your branch.  Only include the commits that provide the
changes.

On Fri, Jun 8, 2018 at 12:05 PM ghronkrepps <notifications@github.com>
wrote:

> What's this PR do?
>
> Add basic Sparkline support.
> Where should the reviewer start?
>
> The Sparklines folder
> How should this be manually tested?
>
> See the Sparklines example
> Any background context you want to provide?
>
> #454 <https://github.com/ClosedXML/ClosedXML/issues/454>
> Screenshots (if appropriate) Questions:
>
>    -
>
>    Is there a blog post? No
>    -
>
>    Does the knowledge base need an update? Yes
>    -
>
>    Does this add new (C#) dependencies? No
>    -
>
>    C# Code Review: @csreviewer
>    -
>
>    Test Automation Review: @csreviewer
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ClosedXML/ClosedXML/pull/907
> Commit Summary
>
>    - Bump version to 0.94
>    - Correctly unprotect worksheet without password
>    - Correctly determine used range when there are merged ranges on a
>    worksheet (fix #881) (#887)
>    - Make calculation engine exception public
>    - Basic sparkline support
>
> File Changes
>
>    - *M* ClosedXML/ClosedXML.csproj
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-0> (2)
>    - *M* ClosedXML/Excel/CalcEngine/Exceptions/CalcEngineException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-1> (2)
>    - *M* ClosedXML/Excel/CalcEngine/Exceptions/CellReferenceException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-2> (8)
>    - *M* ClosedXML/Excel/CalcEngine/Exceptions/CellValueException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-3> (8)
>    - *M* ClosedXML/Excel/CalcEngine/Exceptions/DivisionByZeroException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-4> (10)
>    - *M*
>    ClosedXML/Excel/CalcEngine/Exceptions/NameNotRecognizedException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-5> (8)
>    - *M*
>    ClosedXML/Excel/CalcEngine/Exceptions/NoValueAvailableException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-6> (10)
>    - *M* ClosedXML/Excel/CalcEngine/Exceptions/NullValueException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-7> (8)
>    - *M* ClosedXML/Excel/CalcEngine/Exceptions/NumberException.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-8> (8)
>    - *M* ClosedXML/Excel/EnumConverter.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-9> (79)
>    - *M* ClosedXML/Excel/IXLWorksheet.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-10> (2)
>    - *M* ClosedXML/Excel/Ranges/XLRangeBase.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-11> (36)
>    - *A* ClosedXML/Excel/Sparkline/IXLSparkLineGroup.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-12> (98)
>    - *A* ClosedXML/Excel/Sparkline/IXLSparkLineGroups.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-13> (16)
>    - *A* ClosedXML/Excel/Sparkline/IXLSparkline.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-14> (17)
>    - *A* ClosedXML/Excel/Sparkline/XLSparkLine.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-15> (22)
>    - *A* ClosedXML/Excel/Sparkline/XLSparkLineGroup.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-16> (204)
>    - *A* ClosedXML/Excel/Sparkline/XLSparkLineGroups.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-17> (36)
>    - *M* ClosedXML/Excel/XLWorkbook_Load.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-18> (63)
>    - *M* ClosedXML/Excel/XLWorkbook_Save.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-19> (83)
>    - *M* ClosedXML/Excel/XLWorksheet.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-20> (16)
>    - *M* ClosedXML_Examples/ClosedXML_Examples.csproj
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-21> (2)
>    - *M* ClosedXML_Examples/Creating/CreateFiles.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-22> (1)
>    - *A* ClosedXML_Examples/Sparklines/Sparklines.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-23> (25)
>    - *M* ClosedXML_Sandbox/ClosedXML_Sandbox.csproj
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-24> (2)
>    - *M* ClosedXML_Tests/ClosedXML_Tests.csproj
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-25> (2)
>    - *M* ClosedXML_Tests/Excel/Ranges/UsedAndUnusedCellsTests.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-26> (66)
>    - *A* ClosedXML_Tests/Excel/Worksheets/XLSheetProtectionTests.cs
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-27> (40)
>    - *A* ClosedXML_Tests/Resource/Misc/SHA512PasswordProtection.xlsx
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-28> (0)
>    - *M* appveyor.yml
>    <https://github.com/ClosedXML/ClosedXML/pull/907/files#diff-29> (2)
>
> Patch Links:
>
>    - https://github.com/ClosedXML/ClosedXML/pull/907.patch
>    - https://github.com/ClosedXML/ClosedXML/pull/907.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ClosedXML/ClosedXML/pull/907>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsRjLJXYbyO7e-MUdUqAv2M7GU3z5ks5t6qC2gaJpZM4Ugkmj>
> .
>

vbjay commented:
![image](https://user-images.githubusercontent.com/3629489/41169911-6bfbd03c-6b18-11e8-81e4-16435985f405.png)

ghronkrepps commented:
I am not familiar with how to do that. Is this what I need to do? https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

igitur commented:
> I am not familiar with how to do that. Is this what I need to do? https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Yes, that's what you need to do. But rebase on the `develop` branch, not the `master` branch.

ghronkrepps commented:
Thanks guys. I got it.

Pankraty commented:
From the quick glance, looks quite well!

A few suggestions
1. Inconsistent naming. In ClosedXML collection classes are mostly names just in a plural form. E.g. `XLConditionalFormat` - `XLConditionalFormats`. You have `XLSparkline` - `XLSparklineGroups`. Besides, sometimes you use cameCase (sparkLine).
2. If I am not mistaken, `ReferenceSequence ` aims to store the collection of range addresses to which the sparklines are applied. If it is so I'd suggest using `IXLRanges` instead. A few overloaded methods accepting string addressed, or `IXLRange`, or `IXLRanges` would be nice to have. It brings a few benefits (aside from the consistency with other classes): ranges will be shifted automatically reducing the amount of hard work, and searching for cells having sparklines will become fast and easy.
3. For `Formula` property you may try to utilize the existing `XLFormula` class. For now, they are not shifted automatically, but I hope it will change soon (I'm desperately waiting for the vacation to start with #893 :))
4. Methods accepting the predicate should better define a default value as well (`Predicate<IXLSparkline> predicate = null`) which must be equivalent to `() => true`.

ghronkrepps commented:
@Pankraty Thank you!

1: I followed the XML structure as there are SparklineGroups with SparklineGroup and then inside SparklineGroup there are Sparklines with Sparkline. See this image here:

![image](https://user-images.githubusercontent.com/5125907/41175771-2927decc-6b2c-11e8-8cf5-73a109575665.png)

Is that okay? Also I noticed after getting through some of this that excel calls them "Sparkline" and not "SparkLine" so I tried to change all my references to "Sparkline" I will change any cases I missed.

2, 3, & 4: Sounds good I will work on this.

Pankraty commented:
> SparklineGroups with SparklineGroup and then inside SparklineGroup there are Sparklines with Sparkline

I'll get back to it tomorrow.

vbjay commented:
For something like this. Just go ahead and fixup the latest commit. Do an interactive rebase on develop and change this commit to f instead of pick.   Then push force with lease.  This will roll this commit into the previous one. 
•••

vbjay commented:
Since you are still working on this, change the title to add WIP at the beginning.

ghronkrepps commented:
I'm having issues debugging the project... objects are "null" until I view them in the watch window then they magically have values that I know the have. This is causing problems because the code is literally skipping lines because it thinks the value is null. Never had a problem like this in visual studio before. Anyone know what is going on? I tried cleaning/rebuilding/restarting PC already.

Pankraty commented:
Can this be that you are trying to debug in Release configuration? Try switch to Debug and / or turn off the "Optimize code" checkbox.

ghronkrepps commented:
I am in debugging in debug configuration. Also "Optimize code" is unchecked in all projects.

igitur commented:
I've seen that only with Release configuration. Are you on the latest Visual Studio? 15.7.3 I believe?

ghronkrepps commented:
Yes. 15.7.3 community edition though, which shouldn't matter right?

igitur commented:
Yes, that's what I use at home too. Did you wrote your own console app which references ClosedXML or did you modify the sandbox project?

ghronkrepps commented:
I added an example in the example project and I am running that to do my testing.

igitur commented:
Try modify the `PerformanceRunner.OpenTestFile` method and see if you get the same behaviour. I use that method to do all my initial testing. At least we'll know whether the problem is project specific.

ghronkrepps commented:
I'm having the same results when I try it from the performance runner method.
I'm going to try this again at home and see if the same thing is happening.

This is preventing me from properly deleting sparklines from the spreadsheet. :(
Before
![image](https://user-images.githubusercontent.com/5125907/41376908-8d8aca00-6f28-11e8-9a06-f58f5d8499b4.png)

After expanding "this" in the debug autos window
![image](https://user-images.githubusercontent.com/5125907/41377059-ed00e690-6f28-11e8-954e-a607f2da3896.png)

igitur commented:
Your pdb files could be out of sync. Try deleting all the bin and obj folders and doing a full rebuild of the solution.

Pankraty commented:
Where is the `_sparkline` field actually initialized? I suspect this happens in `Sparkline` getter. This way, when you expand `this` in the watch window you get this property evaluated and the field becomes initialized.

And I have two suggestions regarding this piece of code:
1. Try to avoid storing a sparkline in the `XLCell` field - it makes the class 4 bytes heavier, and since instances of this class are often quite numerous it adds up a significant amount to total memory consumption. Better fetch a sparkline in the property getter - it'll be slower but sparklines are not often used (way rarer than, say, formulas) so I think it's fair.
2. `if (!sparklineGroup.Any()) sparklineGroup.Worksheet.SparklineGroups.Remove(sprklineGroup);` - I am not sure this is the `XLCell`'s responsibility to take care of empty sparkline groups. Perhaps, this can be done in `XLSparklineGroup` class, or, alternatively, empty groups may be skipped during saving.

ghronkrepps commented:
How will I access the sparkline to remove it from the group when the cell is deleted?

Pankraty commented:
It depends on how the data is organized. `XLSparklineGroup` may store the `Dictionary<XLCell, XLSparkline>` and you'll simply remove the item by key. Or, you can find the sparkline right before the cell is actually deleted. Maybe I miss some pitfalls but it should not be hard to make.

ghronkrepps commented:
Thanks. I'll try and get that working tomorrow. Besides that do you see anything else that I missed. Besides deleting everything else seems to be working fine.

Pankraty commented:
Sorry, I missed that you had pushed your changes so I looked at the screenshot only :). Will inspect more carefully.

Pankraty commented on ClosedXML/Excel/Sparkline/XLSparkLineGroup.cs:
```diff
@@ -0,0 +1,263 @@
+            this.Worksheet = worksheet ?? throw new ArgumentNullException(nameof(worksheet));
+
+            AxisColor = XLColor.FromHtml("FF000000");
+            SeriesColor = XLColor.FromHtml("FF376092");
```
I am surprised these are not theme colors but it seems valid.
I would only suggest using `XLColor.Black` instead of `XLColor.FromHtml("FF000000")`

ghronkrepps replied:
I will modify this so the default colors that are black are loaded from XLColor.black.

Pankraty commented on ClosedXML/Excel/Sparkline/XLSparkLine.cs:
```diff
@@ -0,0 +1,36 @@
+    {        
+        public IXLSparklineGroup SparklineGroup { get; }
+
+        public XLFormula Formula { get; set; }
```
I think we should store the reference to the data source range (`IXLRange DataSourceRange`) too. This will make the shifting easy. The only usecase when this is not enough is when the source is defined as the named range. @igitur maybe `IXLNamedRange` must inherit `IXLRange` too?

Pankraty commented on ClosedXML_Examples/Sparklines/Sparklines.cs:
```diff
@@ -0,0 +1,128 @@
+            workbook.SaveAs(filePath);
+        }
+    }    
+}
```
I'm curious how can you go without the unit tests?? =-)

ghronkrepps replied:
I'm not sure what you mean by this. ;)

Pankraty replied:
I mean, you implement the advanced functionality and do not use unit tests for some reason... You do realize that they have to be created anyway, don't you? So why not implement them in advance?

ghronkrepps replied:
I guess because I was unsure my initial attempt at implementing these things was going to be correct so I would just have to rewrite the tests after fixing.

igitur commented on ClosedXML/Excel/Cells/XLCell.cs:
```diff
@@ -176,6 +177,20 @@ internal XLComment Comment
+            {
+                if (_sparkline == null)
+                {
+                    IXLSparklineGroup slg = this.Worksheet.SparklineGroups.Add(this.Worksheet);
```
Well spotted, old chap!

ghronkrepps commented:
Thanks for the generous review. I have fixed the main issue from last night and I think this will solve some of your suggestions you just added. I am going to update the pull request with these changes, and then will go through the list and make updates per your recent suggestions.

Pankraty commented:
OK, great! I will only ask you not to squash your changes until it's ready for merge - this way it's easier to review.

igitur commented:
I want to add my review, but will have time only next week.

Pankraty commented:
I also noticed that some of the tests became terribly slow: https://ci.appveyor.com/project/ClosedXML/closedxml/build/0.94.999.1277/job/nt254cicmg8fa5ho/tests
This has to be fixed too.

Compare with this: https://ci.appveyor.com/project/ClosedXML/closedxml/build/0.94.999.1282/job/7yax0lx4kg81pspp/tests - only 4 tests took longer than a second

ghronkrepps commented on ClosedXML/Excel/Sparkline/IXLSparkLineGroup.cs:
```diff
@@ -0,0 +1,103 @@
+        Boolean DisplayHidden { get; set; }
+        Boolean RightToLeft { get; set; }
+        Double ManualMax { get; set; }
+        Double ManualMin { get; set; }
```
I was just following the patterns as other similar objects in the project I did not see other nullable properties. Can you explain why they should be nullable?

Pankraty replied:
`manualMax` and `manualMin` are not obligatory and only make sense when the corresponding axisType is `custom`. What values are you going to store in these fields when the axis types are auto or group? In my opinion, null would be the best choice. That's why I suggested making them nullable.

ghronkrepps replied:
This makes sense. I will update it.

ghronkrepps commented on ClosedXML/Excel/EnumConverter.cs:
```diff
@@ -780,6 +781,45 @@ public static Xdr.EditAsValues ToOpenXml(this XLPicturePlacement value)
+                case XLSparklineAxisMinMax.Custom: return OfficeExcel.SparklineAxisMinMaxValues.Custom;
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(value), "Not implemented value!");
```
For the enum conversions  I was following the pattern that the other enums were using. It seems like it would this be better to change this when all of the others are changed as well?

Pankraty replied:
Fair point, I missed this. Let's keep in line with other enums

ghronkrepps commented on ClosedXML/Excel/Sparkline/XLSparkLineGroup.cs:
```diff
@@ -0,0 +1,263 @@
+            LowMarkerColor = XLColor.FromColor(sparklineGroup.LowMarkerColor.Color);
+            FirstMarkerColor = XLColor.FromColor(sparklineGroup.FirstMarkerColor.Color);
+            LastMarkerColor = XLColor.FromColor(sparklineGroup.LastMarkerColor.Color);
+            NegativeColor = XLColor.FromColor(sparklineGroup.NegativeColor.Color);
```
I think this is in the CopyFrom method? If I assign the NegativeColor to the NegativeColor would it not be the same XLColor object as the object we are copying from? It's meant to be different so that you can change one without changing the other.

Pankraty replied:
XLColor cannot be changed by it's own. Moreover, each occurence of the same color (say, black) is actually the same instance. By direct assigning you save time by not searching the color in the repository by its key.
And when I asked about NegativeColor I meant all the color properties in this section, of course.

vbjay commented:
Enum constraints and enum traits.
https://blogs.msdn.microsoft.com/seteplia/2018/06/12/dissecting-new-generics-constraints-in-c-7-3/

On Fri, Jun 15, 2018, 10:45 AM Aleksei <notifications@github.com> wrote:

> *@Pankraty* commented on this pull request.
> ------------------------------
>
> In ClosedXML/Excel/Sparkline/IXLSparkLineGroup.cs
> <https://github.com/ClosedXML/ClosedXML/pull/907#discussion_r195764041>:
>
> > +        XLColor SeriesColor { get; set; }
> +        XLColor NegativeColor { get; set; }
> +        XLColor MarkersColor { get; set; }
> +
> +        Boolean Markers { get; set; }
> +        Boolean High { get; set; }
> +        Boolean Low { get; set; }
> +        Boolean First { get; set; }
> +        Boolean Last { get; set; }
> +        Boolean Negative { get; set; }
> +        Boolean DateAxis { get; set; }
> +        Boolean DisplayXAxis { get; set; }
> +        Boolean DisplayHidden { get; set; }
> +        Boolean RightToLeft { get; set; }
> +        Double ManualMax { get; set; }
> +        Double ManualMin { get; set; }
>
> manualMax and manualMin are not obligatory and only make sense when the
> corresponding axisType is custom. What values are you going to store in
> these fields when the axis types are auto or group? In my opinion, null
> would be the best choice. That's why I suggested making them nullable.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ClosedXML/ClosedXML/pull/907#discussion_r195764041>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsYBCrWl-x95NzQAojkeg6lzNyVdIks5t88ibgaJpZM4Ugkmj>
> .
>

ghronkrepps commented:
So I believe I fixed the issue with the slow/failed tests. It had to do with deleting the sparklines when the range was deleted.

I also added an example test for Sparklines -- I think that is how you wanted me to do the test?
An issue with my testing though that is confusing me. I noticed when the the XLWorkbook_Save.CreateParts() was running during the test it was loading in a saved workbook file, and then using the worksheet extension lists from that. I had to delete the sparkline groups from this before adding mine, otherwise they were duplicating. I guess it's necessary to delete those parts?

Also if you can take a look at the test there are some very minor differences that is causing the test to fail and I can not figure out how/why it is changing the output. For example:

only the span is different?
![image](https://user-images.githubusercontent.com/5125907/41488802-a0ac6b5e-70bb-11e8-92c7-0b3293308e4c.png)

its shifting the row order around (data is staying the same)
![image](https://user-images.githubusercontent.com/5125907/41488873-dd3679c0-70bb-11e8-8695-3cbb8839557d.png)
![image](https://user-images.githubusercontent.com/5125907/41488881-e67ca7de-70bb-11e8-8c7d-c14b7ae967cb.png)
![image](https://user-images.githubusercontent.com/5125907/41488891-efa6de06-70bb-11e8-953e-dd54f069bdcf.png)

Pankraty commented:
The test you've added is fine but it is far not enough. Take a look into this comment (https://github.com/ClosedXML/ClosedXML/issues/454#issuecomment-395031293). There was a rough list of thing to be tested - each bullet stands for a separate test case or even for multiple ones. And I am sure this list is not complete and can be extended even more.

As for duplicating items on workbook saving - yes, existing entries are not automatically deleted, you have to perform it manually. The simplest solution would be to remove the entire section and then to generate all the sparklines from scratch. The drawback of this approach is that there is a risk to lose some data (e.g. if one day Excel will start to save some additional attributes to XML we may miss this, thus these attributes won't be read so they won't be saved as well). A possible solution would be to preserve the original number of the sparkline group as has been read from the file, and then during saving to update the modified entries, to remove deleted ones and to add the new ones. It is harder, of course, but still doable.

I don't know the reasons for the issues with the test you described (items reordering, changed `spans`, etc.) and unfortunately don't have time right now to investigate, but will try to go back to it later this weekend.

ghronkrepps commented:
I have just submitted some more changes to expand the tests and add some functionality with the copying  and deleting of sparklines. There remains one issue that I cannot figure out in the testing that I described in my previous comment. I have spent several hours trying to figure it out but I believe it is just beyond my current knowledge of this library to find the true cause. I have narrowed it down to the XLRangeBase.Clear method and isolated it into a test called "DeletingSparklinesHelp". Can you please review that test and help me discover the cause?

I believe the majority of sparkline functionality is available with this pull request. I know the formulas will not shift/update when rows or columns are added or deleted. I would like to create a different issue for that functionality (it looks to be way beyond my current understanding, and I believe that formula updating is not even implemented yet for conditional formatting/data validations.)

Then we can get this current pull request approved.

Pankraty commented:
There are still lots of things to do before this can be merged into the `develop` branch. Besides, @igitur has not provided his feedback yet (and I believe he has _many_ suggestions ;) 
On the other hand, I do understand that this task may be too giant for being your very first contribution to ClosedXML. I am willing to help you with it, but right now I don't have time at all. I am really sorry about that.
Good news is that my vacation starts in a week and I'll be able to invest my effort into this task then.

igitur commented:
Feedback given. I'm sorry... I previously said that sparklines are a subset of charts. I see now I'm completely wrong. Don't know what gave me that idea. When you logged the issue initially, I created some sparklines in Excel, saved the file and inspected the contents. I'm pretty sure I saw some chart objects in there, which is what made me park the issue until we were going to implement charts fully.

Anyway, this is great work, especially for a first PR. Congrats. @Pankraty is the master of computing patterns and I try to make the API as intuitive for the end-user as possible, so most of my feedback is about the properties and methods that are exposed.

ghronkrepps commented:
@igitur No problem. I had done some initial research into the sparklines and saw they were not charts, but rather their own thing. I have looked over your comments but will go more in depth and fix/respond to them sometime this week.

Pankraty commented:
I've started with sparklines refactoring. It is far from finishing but you can see the API changes here:
https://github.com/Pankraty/ClosedXML/commit/8a6b5a9822706f946df20e4d92f5d75d15176c66

igitur commented:
@Pankraty Are you taking this issue over then?

Pankraty commented:
Yes, I suppose so.

Pankraty commented on ClosedXML/Excel/XLWorkbook_Load.cs:
```diff
@@ -2363,6 +2363,62 @@ private void LoadExtensions(WorksheetExtensionList extensions, XLWorksheet ws)
+                    if (slg.High != null) sparklineGroup.High = slg.High;
+                    if (slg.Low != null) sparklineGroup.Low = slg.Low;                                        
+                    if (slg.First != null) sparklineGroup.Low = slg.First;
+                    if (slg.Last != null) sparklineGroup.Low = slg.Last;
```
```
                    if (slg.Low != null) sparklineGroup.Low = slg.Low;                                        
                    if (slg.First != null) sparklineGroup.Low = slg.First;
                    if (slg.Last != null) sparklineGroup.Low = slg.Last;
```

That's why I like unit tests

igitur replied:
I wouldn't say I like them, but that's why they're a necessity.

Pankraty commented on ClosedXML/Excel/Cells/IXLCell.cs:
```diff
@@ -366,6 +366,7 @@ public interface IXLCell
 
         IXLRow WorksheetRow();
 
+        Boolean HasSparkline { get; }
```
Done

Pankraty commented:
Basically, the feature is ready for review.

**TODO**
* Add a few more tests (TODOs in code)
* Add reference files to make sure the file is saved correctly and clean up Examples project
* Fix saving (depends on #949). This will fix a failing test too
* Add documentation in code
* Add documentation in Wiki

ghronkrepps commented:
@Pankraty I am using your updated build of the sparklines and I no longer see a way to add an empty sparkline group to the worksheet, as it is not available in the public API. Is this your intention? If a user (me in this case) wants to add an empty sparkline group and then add sparklines to it elsewhere in the project but I am unable to do so now.

Pankraty commented:
> I noticed the file names some where "SparkLine" with the capital "L". I tried to fix this a couple times to "Sparkline" but they kept reverting to "SparkLine"... maybe you can fix that?

I noticed this too. Git does not even consider these renamed files as being modified. Perhaps, because we both work on Windows. Maybe on Linux it'll work differently. Or, I can try to rename those files to something different, commit and then rename back with the proper case.

>  I no longer see a way to add an empty sparkline group to the worksheet, as it is not available in the public API

Well, you cannot do this in Excel either, can you? I guess I could add the possibility to do this, it just did not look like a common use case to me.

ghronkrepps commented:
> Well, you cannot do this in Excel either, can you? I guess I could add the possibility to do this, it just did not look like a common use case to me.

Yes, I guess that's true.

Pankraty commented:
> It would just be to avoid code like this:...

In case you don't care about existing sparklines and want to create them again you can use `slg = ws.SparklineGroups.Add(trendRange, trendDataRange);` and they will be overwritten (at least, it is supposed to be so). If, on the contrary, you do care about the existing ones you would have to process them separately even if it was possible to create empty sparkline groups.

igitur commented:
Closed by #1175
